R/utils-latency.R

```diff
@@ -0,0 +1,294 @@
+    # get the as_of in an `adjust_latency` step, regardless of where
+    handpicked_forecast_date <- map(
+      this_recipe$steps,
+      function(x) {
+        if (inherits(x, "step_adjust_latency")) x$forecast_date
+      }
+    ) %>% Filter(Negate(is.null), .)
+    if (length(handpicked_forecast_date) > 0) {
+      max_time_value <- handpicked_forecast_date[[1]]
+    } else {
+      # if we haven't chosen one, use either the max_time_value or the as_of
+      max_time_value <- max(
+        max_time_value,
+        attributes(new_data)$metadata$as_of
+      )
+    }
+  }
+  max_time_value
+}
+
+
+#' pad every group at the right interval
+#' @description
+#' Perform last observation carried forward on a group by group basis. It uses
+#'   `guess_period` to find the appropriate interval to fill-forward by. It
+#'   maintains the grouping structure it recieves. It does *not* fill any
+#'   "interior" `NA` values occurring in the data beforehand.
+#' @param x an epi_df to be filled forward.
+#' @param columns_to_complete which columns to apply completion to. By default every non-key column of an epi_df
+#' @param groups the grouping by which to fill forward
+#' @importFrom tidyselect all_of
+#' @importFrom rlang list2
+#' @importFrom vctrs vec_cast
+#' @importFrom dplyr across arrange bind_rows group_by summarise
+#' @keywords internal
+pad_to_end <- function(x, groups, end_date, columns_to_complete = NULL) {
+  if (is.null(columns_to_complete)) {
+    columns_to_complete <- setdiff(names(x), key_colnames(x))
+  }
+  itval <- epiprocess:::guess_period(c(x$time_value, end_date), "time_value")
+  # get the time values we need to fill in
+  completed_time_values <-
+    x %>%
+    group_by(across(all_of(groups))) %>%
+    summarise(
+      time_value = list2(
+        time_value = seq_null_swap(from = max(time_value) + itval, to = end_date, by = itval)
+      )
+    ) %>%
+    unnest("time_value") %>%
+    mutate(time_value = vec_cast(time_value, x$time_value))
+  # pull the last value in each group and fill forward
+  grouped_and_arranged <- x %>%
+    arrange(across(all_of(c("time_value", groups)))) %>%
+    group_by(across(all_of(groups)))
+  values_to_fill <- grouped_and_arranged %>%
+    slice(min(across(all_of(columns_to_complete), count_single_column)):n())
+  filled_values <- values_to_fill %>%
+    bind_rows(completed_time_values) %>%
+    arrange(across(all_of(c("time_value", groups)))) %>%
+    fill(all_of(columns_to_complete), .direction = "down") %>%
+    slice(-1) # remove the oirginal rows
+
+  grouped_and_arranged %>%
+    slice(1:min(across(all_of(columns_to_complete), count_single_column))) %>%
+    bind_rows(filled_values) %>%
+    arrange(across(all_of(key_colnames(x)))) %>%
+    ungroup() %>%
+    group_by(across(all_of(get_grouping_columns(x))))
+}
+
+#' return the names of the grouped columns, or `NULL`
+#' @param x an epi_df
+#' @keywords internal
+#' @importFrom utils head
+get_grouping_columns <- function(x) {
```

Comment:
This seems to be used in only one location. Where would the `groups` attr be set? Is this different from `key_colnames()`?

Reply:
if `x` is already grouped, this means the step preserves any existing group structure, regardless of how it's grouped. 
```
a <- tibble(
+   geo_value = rep("place", 200),
+   time_value = as.Date("2021-01-01") + 0:199,
+   case_rate = sqrt(1:200) + atan(0.1 * 1:200) + sin(5 * 1:200) + 1,
+   death_rate = atan(0.1 * 1:200) + cos(5 * 1:200) + 1
+ ) %>% group_by(geo_value)
+ attributes(a)$groups
# A tibble: 1 × 2
  geo_value       .rows
  <chr>     <list<int>>
1 place           [200]
```
if `x` is ungrouped, this is just `null` so grouping by the result is a no-op.

Its a function not really for reuse but to reduce the overhead of reading `pad_to_end`, because the way of getting the previous grouping is a bit jank.